### ISSUE TITLE
how to edit dd.mm.yyyy dateformat in  apache superset-ui-chart-controls

### ISSUE THREAD

**Issue by mujohiddin** [apache/superset#28615]
[###]([url](![photo_2024-05-21_12-22-29](https://github.com/apache/superset/assets/69301931/4c5c53ba-e2eb-40db-9c92-bb3474773b90))) Bug description

26 countries use "dd.mm.yyyy" as date format. (https://en.wikipedia.org/wiki/Date_format_by_country)

I'd like to add this format ("%d.%m.%Y "in D3 notation) below "Adaptive Formatting" in
![162387285-634840d8-6bd6-4e19-9361-fbccef0c58c9](https://github.com/apache/superset/assets/69301931/70e57450-9f43-4d48-aa8d-6868d4dda5d3)
 but not working
![162386896-2d85711f-0550-40b9-a9d6-dd247bb8e2fa](https://github.com/apache/superset/assets/69301931/c8daf5f6-1ce1-47ee-89fb-1ca6d38a04b6)

### How to reproduce the bug

how to edit datetime

### Superset version

master / latest-dev

### Python version

3.9

### Node version

16

### Browser

Chrome

### Checklist

- [ ] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [ ] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [ ] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by mujohiddin** [comment]
![photo_2024-05-21_12-22-29](https://github.com/apache/superset/assets/69301931/e6355c94-6a88-4845-9ae4-9e2879374a4a)


In this case
options: %d.%m.%Y, "%d/%m/%Y" I entered like this

**Comment by mujohiddin** [comment]
![photo_2024-05-21_12-25-36](https://github.com/apache/superset/assets/69301931/42b922ea-c235-4c84-afb7-6e877b61ec6a)
![photo_2024-05-21_12-25-04](https://github.com/apache/superset/assets/69301931/3860dee0-e12f-4b79-897e-94808bb6e2f7)

In this case
options: %d.%m.%Y, "%d/%m/%Y" I entered like this

**Comment by mujohiddin** [comment]
In this case
options: %d.%m.%Y, "%d/%m/%Y" I entered like this
![photo_2024-05-21_12-25-36](https://github.com/apache/superset/assets/69301931/3b470d4c-267d-42c6-8637-d021d4cafa19)
![Uploading photo_2024-05-21_12-27-17.jpg…]()

**Comment by mujohiddin** [comment]
![Uploading photo_2024-05-21_12-27-17.jpg…]()

**Comment by mujohiddin** [comment]
![Uploading photo_2024-05-21_12-28-3764546.jpg…]()

**Comment by mujohiddin** [comment]
![Uploading photo_2024-05-21_12-28-3764546.jpg…]()

**Comment by maximStafeevAtAllscripts** [comment]
Hi! I have the same issue. I need to show all my timestamps in %Y-%m-%d %H:%M:%S.%L format with thousands of second but cannot change format in UI. I updated chart via API with this format and it works, but I wand to have it in UI also. 
I found all this array of D3 format options in D3Formatting.ts. It seems it is hardcoded.

**Comment by maximStafeevAtAllscripts** [comment]
@dosu but I'm using docker image. I do not build superset from sources. I cannot make changes in UI that way.

**Comment by sfirke** [comment]
@mujohiddin @maximStafeevAtAllscripts am I understanding this correctly that the core problem here is that we can't type or copy-paste into that field?  We should be able to put any valid D3 string in that field but despite being formatted like a field that accepts text, it functions as a dropdown.

I'm experiencing this in 4.0.2.  If you agree we are talking about the same issue, I will re-write this one to clarify it.

**Comment by goldjee** [comment]
> we can't type or copy-paste into that field?

I have a similar case, and for me, this is the issue. Also, I've found out that we can define default D3 formatting for numbers in `superset_config.py`, like so:

```Python
D3_FORMAT = {
    "decimal": ".",
    "thousands": " ",
    "grouping": [3],
}
```

Is there a similar option to define default formatting for dates?

**Comment by maximStafeevAtAllscripts** [comment]
@sfirke yes, accepting any valid D3 string in that field in the UI would be a solution. Also why not allow extending that predefined array in superset_config.py?

**Comment by President** [comment]
This limitation is quite frustrating. It’s unclear why users are restricted to selecting from a set of predefined formats, especially when these formats don’t align with the date conventions of many countries. It would be much more flexible to allow custom date formats, enabling users to specify the exact format they need.

In countries with non-standard formats, such as DD.MM.YYYY or similar, this restriction prevents compliance with local conventions and adds unnecessary friction. Could we explore adding a free-form date format option here?

<img width="396" alt="image" src="https://github.com/user-attachments/assets/c24a3503-93dd-4064-be62-64484dfa18ff">

**Comment by sfirke** [comment]
Yes I think everyone agrees it would be nice to have open text allowed here like we do for number formatting. I don't think it's a design decision, more of an oversight.

It's just a matter of someone doing it. If anyone feels able to take this on, please go for it and open a PR.

**Comment by rusackas** [comment]
[This PR](https://github.com/apache/superset/pull/32814) seems to have solved this one!
